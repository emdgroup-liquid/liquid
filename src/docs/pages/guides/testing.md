---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

