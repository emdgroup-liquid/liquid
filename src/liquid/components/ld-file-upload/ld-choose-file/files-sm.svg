<svg width="89" height="76" viewBox="0 0 89 76" fill="none" xmlns="http://www.w3.org/2000/svg">
<g clip-path="url(#clip0_353_683)">
<rect width="33.1967" height="43.4171" transform="matrix(0.97058 -0.24078 0.245017 0.969519 0.191895 33.1506)" fill="#2DBECD"/>
<path d="M17.2248 66.7345L27.9918 64.0634C28.4863 63.9408 28.9734 64.2408 29.1012 64.7468C29.2291 65.2528 28.9422 65.7448 28.4477 65.8674L17.6807 68.5385C17.1862 68.6612 16.6991 68.3612 16.5712 67.8552C16.4433 67.3492 16.7303 66.8572 17.2248 66.7345Z" fill="white" fill-opacity="0.8"/>
<path d="M16.4493 63.6256L32.5698 59.6265C33.069 59.5027 33.5597 59.8018 33.6875 60.3078C33.8154 60.8138 33.5249 61.3067 33.0257 61.4305L16.9053 65.4297C16.4061 65.5535 15.9154 65.2544 15.7875 64.7484C15.6596 64.2424 15.9502 63.7495 16.4493 63.6256Z" fill="white" fill-opacity="0.8"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M24.7585 43.0322C24.8315 43.321 24.6472 43.6155 24.347 43.69L20.9493 44.5329C20.649 44.6074 20.3465 44.4337 20.2735 44.1449C20.2005 43.8561 20.3847 43.5616 20.685 43.4871L24.0827 42.6442C24.3829 42.5697 24.6855 42.7435 24.7585 43.0322Z" fill="white" fill-opacity="0.8"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M26.0196 57.1672C25.1189 57.3906 24.2112 56.8694 23.9922 56.003L21.085 44.4992L24.3468 43.69L27.254 55.1939C27.473 56.0602 26.9203 56.9437 26.0196 57.1672Z" fill="white" fill-opacity="0.8"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M26.0196 57.1672C25.1189 57.3906 24.2112 56.8694 23.9922 56.003L23.1002 52.4734L26.362 51.6643L27.254 55.1939C27.473 56.0602 26.9203 56.9437 26.0196 57.1672Z" fill="#008490"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M19.8654 44.2459C19.9384 44.5347 19.7542 44.8291 19.4539 44.9036L16.0562 45.7465C15.756 45.821 15.4534 45.6473 15.3804 45.3585C15.3075 45.0697 15.4917 44.7752 15.7919 44.7007L19.1896 43.8578C19.4899 43.7833 19.7924 43.9571 19.8654 44.2459Z" fill="white" fill-opacity="0.8"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M21.127 58.3808C20.2263 58.6042 19.3186 58.083 19.0996 57.2167L16.1924 45.7128L19.4542 44.9036L22.3615 56.4075C22.5804 57.2739 22.0277 58.1573 21.127 58.3808Z" fill="white" fill-opacity="0.8"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M21.127 58.3808C20.2263 58.6042 19.3186 58.083 19.0996 57.2167L16.8862 48.458L20.148 47.6489L22.3615 56.4075C22.5804 57.2739 22.0277 58.1573 21.127 58.3808Z" fill="#008490"/>
<rect opacity="0.2" width="27.6801" height="36.5734" transform="matrix(1.00094 0.0167532 -0.00854105 0.998894 17.3301 18.0354)" fill="#0F69AF"/>
</g>
<path d="M56.4775 18.9743L80.726 25.5119L86.3079 35.6878L77.0252 69.4856L44.9776 60.8453L56.4775 18.9743Z" fill="#FFC832"/>
<path d="M81.0049 25.5866L86.2782 35.7998L78.7568 33.7719L81.0049 25.5866Z" fill="white" fill-opacity="0.8"/>
<rect width="1.75069" height="11.1655" rx="0.875345" transform="matrix(0.965524 0.260314 -0.264845 0.964291 60.5151 36.7665)" fill="white" fill-opacity="0.8"/>
<rect width="1.75069" height="11.1655" rx="0.875345" transform="matrix(0.965524 0.260314 -0.264845 0.964291 63.1445 37.4752)" fill="white" fill-opacity="0.8"/>
<rect width="1.75069" height="11.1655" rx="0.875345" transform="matrix(0.965524 0.260314 -0.264845 0.964291 65.7739 38.1841)" fill="white" fill-opacity="0.8"/>
<rect width="1.75069" height="11.1655" rx="0.875345" transform="matrix(0.965524 0.260314 -0.264845 0.964291 68.4033 38.893)" fill="white" fill-opacity="0.8"/>
<rect width="1.75069" height="11.1655" rx="0.875345" transform="matrix(0.965524 0.260314 -0.264845 0.964291 71.0327 39.6019)" fill="white" fill-opacity="0.8"/>
<rect width="1.75069" height="11.1655" rx="0.875345" transform="matrix(0.965524 0.260314 -0.264845 0.964291 73.6621 40.3108)" fill="white" fill-opacity="0.8"/>
<rect width="1.75069" height="7.82868" rx="0.875345" transform="matrix(0.965524 0.260314 -0.264845 0.964291 59.6313 39.9839)" fill="#A37303"/>
<rect width="1.75069" height="2.43844" rx="0.875345" transform="matrix(0.965524 0.260314 -0.264845 0.964291 60.8335 45.8905)" fill="#A37303"/>
<rect width="1.75069" height="3.72183" rx="0.875345" transform="matrix(0.965524 0.260314 -0.264845 0.964291 63.8027 45.362)" fill="#A37303"/>
<rect width="1.75069" height="7.82868" rx="0.875345" transform="matrix(0.965524 0.260314 -0.264845 0.964291 67.5195 42.1107)" fill="#A37303"/>
<rect width="1.75069" height="6.48112" rx="0.875345" transform="matrix(0.965524 0.260314 -0.264845 0.964291 69.792 44.119)" fill="#A37303"/>
<rect width="1.75069" height="3.786" rx="0.875345" transform="matrix(0.965524 0.260314 -0.264845 0.964291 71.7075 47.4268)" fill="#A37303"/>
<path d="M28.7119 0.000457764H53.8453L61.9292 8.36612V43.3885H28.7119V0.000457764Z" fill="#0F69AF"/>
<path d="M54.1343 0L61.9302 8.48187H54.1343V0Z" fill="white" fill-opacity="0.8"/>
<path d="M35.4136 11.4175H52.4168C52.9433 11.4175 53.3553 11.814 53.3553 12.3208C53.3553 12.8275 52.9433 13.224 52.4168 13.224H35.4136C34.8871 13.224 34.4751 12.8275 34.4751 12.3208C34.4751 11.814 34.8871 11.4175 35.4136 11.4175Z" fill="white" fill-opacity="0.8"/>
<path d="M35.4126 20.8784H52.4158C52.9423 20.8784 53.3543 21.275 53.3543 21.7817C53.3543 22.2884 52.9423 22.685 52.4158 22.685H35.4126C34.8861 22.685 34.4741 22.2884 34.4741 21.7817C34.4741 21.275 34.8861 20.8784 35.4126 20.8784Z" fill="white" fill-opacity="0.8"/>
<path d="M35.4126 14.6193H46.8763C47.4028 14.6193 47.8148 15.0159 47.8148 15.5226C47.8148 16.0293 47.4028 16.4258 46.8763 16.4258H35.4126C34.8862 16.4258 34.4741 16.0293 34.4741 15.5226C34.4741 15.0159 34.8862 14.6193 35.4126 14.6193Z" fill="white" fill-opacity="0.8"/>
<defs>
<clipPath id="clip0_353_683">
<rect width="33.1967" height="43.4171" fill="white" transform="matrix(0.97058 -0.24078 0.245017 0.969519 0.191895 33.1506)"/>
</clipPath>
</defs>
</svg>
