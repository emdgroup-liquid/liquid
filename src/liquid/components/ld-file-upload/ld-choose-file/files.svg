<svg width="142" height="122" viewBox="0 0 142 122" fill="none" xmlns="http://www.w3.org/2000/svg">
<g clip-path="url(#clip0_333_531)">
<rect width="53" height="70" transform="matrix(0.969948 -0.243312 0.242469 0.970159 0.305664 53.483)" fill="#2DBECD"/>
<path d="M27.4816 107.665L44.6605 103.356C45.4495 103.158 46.2266 103.642 46.4306 104.458C46.6346 105.274 46.1769 106.068 45.3879 106.266L28.209 110.575C27.4201 110.773 26.6429 110.289 26.4389 109.473C26.2349 108.657 26.6927 107.863 27.4816 107.665Z" fill="white" fill-opacity="0.8"/>
<path d="M26.2439 102.649L51.9642 96.1974C52.7605 95.9976 53.5435 96.4802 53.7475 97.2965C53.9515 98.1129 53.488 98.9081 52.6916 99.1079L26.9713 105.56C26.1749 105.76 25.392 105.277 25.1879 104.461C24.9839 103.644 25.4475 102.849 26.2439 102.649Z" fill="white" fill-opacity="0.8"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M39.5011 69.4254C39.6175 69.8913 39.3236 70.3664 38.8445 70.4866L33.4235 71.8464C32.9444 71.9666 32.4617 71.6863 32.3453 71.2204C32.2288 70.7545 32.5228 70.2794 33.0018 70.1592L38.4229 68.7993C38.9019 68.6792 39.3846 68.9594 39.5011 69.4254Z" fill="white" fill-opacity="0.8"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M41.5142 92.2297C40.0771 92.5902 38.6289 91.7493 38.2796 90.3516L33.641 71.792L38.8453 70.4865L43.4838 89.0461C43.8331 90.4438 42.9513 91.8692 41.5142 92.2297Z" fill="white" fill-opacity="0.8"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M41.5142 92.2297C40.0771 92.5902 38.6289 91.7493 38.2796 90.3516L36.8564 84.6571L42.0606 83.3517L43.4838 89.0461C43.8331 90.4438 42.9513 91.8692 41.5142 92.2297Z" fill="#008490"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M31.6944 71.3834C31.8109 71.8493 31.5169 72.3244 31.0379 72.4446L25.6168 73.8044C25.1378 73.9246 24.6551 73.6443 24.5386 73.1784C24.4222 72.7125 24.7161 72.2374 25.1952 72.1172L30.6162 70.7573C31.0953 70.6372 31.578 70.9175 31.6944 71.3834Z" fill="white" fill-opacity="0.8"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M33.7076 94.1877C32.2705 94.5482 30.8223 93.7073 30.4729 92.3096L25.8344 73.75L31.0386 72.4445L35.6772 91.0041C36.0265 92.4018 35.1447 93.8272 33.7076 94.1877Z" fill="white" fill-opacity="0.8"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M33.7076 94.1877C32.2705 94.5482 30.8223 93.7073 30.4729 92.3096L26.9413 78.179L32.1455 76.8735L35.6772 91.0041C36.0265 92.4018 35.1447 93.8272 33.7076 94.1877Z" fill="#008490"/>
<rect opacity="0.2" width="44.2214" height="58.9295" transform="matrix(0.999639 0.0169183 -0.00845752 1.00018 27.6494 29.0971)" fill="#0F69AF"/>
</g>
<path d="M90.1104 30.6119L128.799 41.1592L137.705 57.5763L122.894 112.103L71.7621 98.1637L90.1104 30.6119Z" fill="#FFC832"/>
<path d="M129.244 41.2798L137.658 57.757L125.657 54.4854L129.244 41.2798Z" fill="white" fill-opacity="0.8"/>
<rect width="2.79536" height="17.9998" rx="1.39768" transform="matrix(0.96479 0.263023 -0.262121 0.965035 96.5518 59.3164)" fill="white" fill-opacity="0.8"/>
<rect width="2.79536" height="17.9998" rx="1.39768" transform="matrix(0.96479 0.263023 -0.262121 0.965035 100.747 60.46)" fill="white" fill-opacity="0.8"/>
<rect width="2.79536" height="17.9998" rx="1.39768" transform="matrix(0.96479 0.263023 -0.262121 0.965035 104.942 61.6037)" fill="white" fill-opacity="0.8"/>
<rect width="2.79536" height="17.9998" rx="1.39768" transform="matrix(0.96479 0.263023 -0.262121 0.965035 109.138 62.7474)" fill="white" fill-opacity="0.8"/>
<rect width="2.79536" height="17.9998" rx="1.39768" transform="matrix(0.96479 0.263023 -0.262121 0.965035 113.333 63.8911)" fill="white" fill-opacity="0.8"/>
<rect width="2.79536" height="17.9998" rx="1.39768" transform="matrix(0.96479 0.263023 -0.262121 0.965035 117.528 65.0348)" fill="white" fill-opacity="0.8"/>
<rect width="2.79536" height="12.6205" rx="1.39768" transform="matrix(0.96479 0.263023 -0.262121 0.965035 95.1416 64.5074)" fill="#A37303"/>
<rect width="2.79536" height="3.93099" rx="1.39768" transform="matrix(0.96479 0.263023 -0.262121 0.965035 97.0596 74.0367)" fill="#A37303"/>
<rect width="2.79536" height="5.99993" rx="1.39768" transform="matrix(0.96479 0.263023 -0.262121 0.965035 101.797 73.1841)" fill="#A37303"/>
<rect width="2.79536" height="12.6205" rx="1.39768" transform="matrix(0.96479 0.263023 -0.262121 0.965035 107.728 67.9386)" fill="#A37303"/>
<rect width="2.79536" height="10.4481" rx="1.39768" transform="matrix(0.96479 0.263023 -0.262121 0.965035 111.354 71.1786)" fill="#A37303"/>
<rect width="2.79536" height="6.10337" rx="1.39768" transform="matrix(0.96479 0.263023 -0.262121 0.965035 114.409 76.5153)" fill="#A37303"/>
<path d="M45.8096 0.000732422H85.9101L98.8079 13.4973V70.0001H45.8096V0.000732422Z" fill="#0F69AF"/>
<path d="M86.3711 -3.05176e-05L98.8095 13.684H86.3711V-3.05176e-05Z" fill="white" fill-opacity="0.8"/>
<path d="M56.5023 18.4202H83.631C84.471 18.4202 85.1283 19.06 85.1283 19.8775C85.1283 20.695 84.471 21.3348 83.631 21.3348H56.5023C55.6623 21.3348 55.0049 20.695 55.0049 19.8775C55.0049 19.06 55.6623 18.4202 56.5023 18.4202Z" fill="white" fill-opacity="0.8"/>
<path d="M56.5013 33.6838H83.63C84.47 33.6838 85.1274 34.3236 85.1274 35.1411C85.1274 35.9586 84.47 36.5984 83.63 36.5984H56.5013C55.6613 36.5984 55.0039 35.9586 55.0039 35.1411C55.0039 34.3236 55.6613 33.6838 56.5013 33.6838Z" fill="white" fill-opacity="0.8"/>
<path d="M56.5013 23.5858H74.7916C75.6316 23.5858 76.289 24.2256 76.289 25.0431C76.289 25.8606 75.6316 26.5003 74.7916 26.5003H56.5013C55.6613 26.5003 55.0039 25.8606 55.0039 25.0431C55.0039 24.2256 55.6613 23.5858 56.5013 23.5858Z" fill="white" fill-opacity="0.8"/>
<defs>
<clipPath id="clip0_333_531">
<rect width="53" height="70" fill="white" transform="matrix(0.969948 -0.243312 0.242469 0.970159 0.305664 53.483)"/>
</clipPath>
</defs>
</svg>
