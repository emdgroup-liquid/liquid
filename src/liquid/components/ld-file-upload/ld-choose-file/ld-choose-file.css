:host {
  background: var(--ld-col-wht);
  border-radius: var(--ld-br-l);
  border: var(--ld-sp-2) dashed var(--ld-col-neutral-100);
  box-sizing: border-box;
  list-style: none;
  margin: 0;
  padding: 0;
}

/* :host[highlighted] {
  background: var(--ld-col-rb-050);
} */

:host(.ld-choose-file--highlighted) {
  background: var(--ld-col-rb-050);
}

.ld-choose-file--sm {
  display: flex;
  flex-direction: row;
  gap: var(--ld-sp-40);
  align-items: center;
  margin: var(--ld-sp-16) var(--ld-sp-40) var(--ld-sp-16) var(--ld-sp-40);
  justify-content: center;

  .ld-choose-file__image {
    width: 5.5625rem;
    height: auto;
  }
}

.ld-choose-file--bg {
  display: flex;
  flex-direction: column;
  gap: var(--ld-sp-32);
  align-items: center;
  margin: var(--ld-sp-32) var(--ld-sp-40) var(--ld-sp-32) var(--ld-sp-40);
}

.ld-choose-file__text {
  display: grid;
  grid-template-rows: 1fr 1fr 1fr;
  align-items: center;
  grid-gap: var(--ld-sp-8);
  justify-items: center;
  position: relative;
  text-align: center;
}

/* ::slotted(*) {
  position: absolute;
  inset: 0;
  opacity: 0;
} */

/* ::slotted(input) {
  position: absolute;
  inset: 0;
  opacity: 0;
} */
