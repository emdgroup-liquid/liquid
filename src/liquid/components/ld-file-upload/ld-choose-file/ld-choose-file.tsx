import {
  Component,
  h,
  Host,
  Prop,
  Element,
  State,
  Event,
  EventEmitter,
  Fragment,
} from '@stencil/core'
import { getClassNames } from '../../../utils/getClassNames'
import { getAssetPath } from '../../../utils/assetPath'
import type { UploadItem } from '../ld-file-upload'

/**
 * @virtualProp ref - reference to component
 */
@Component({
  tag: 'ld-choose-file',
  styleUrl: 'ld-choose-file.css',
  shadow: true,
  assetsDirs: ['assets'],
})
export class LdChooseFile {
  @Element() el: HTMLLdChooseFileElement

  /** Max. file size in bytes */
  @Prop() maxSize?: number = 1572864

  /** startUpload defines whether upload starts immediately after choosing files or after confirmation. */
  @Prop() startUpload?: boolean = false

  /** contineClicked defines whether the continue button has been clicked while startUpload = false */
  @Prop() continueClicked?: boolean = false

  /** selectMultiple defines whether selection of multiple input files is allowed. */
  @Prop() selectMultiple?: boolean = false

  /** Chosen Files */
  /* @Prop() uploadFiles: {
    state:
      | 'pending'
      | 'paused'
      | 'cancelled'
      | 'uploading'
      | 'uploaded'
      | 'upload failed'
    fileName: string
    fileSize: number
    fileType: string
    progress: number
  }[] = [] */
  @Prop() uploadFiles: UploadItem[] = []

  /** Chosen Files from the parent component */
  /* @Prop() uploadItems: {
    state:
      | 'pending'
      | 'paused'
      | 'cancelled'
      | 'uploading'
      | 'uploaded'
      | 'upload failed'
    fileName: string
    fileSize: number
    fileType: string
    progress: number
  }[] = [] */
  @Prop() uploadItems: UploadItem[] = []

  /** Size of the choose file area */
  @Prop() size?: 'sm' | 'bg' = 'bg'

  /** Represents whether a file is currently being dragged over the drop area */
  @State() highlighted? = false

  /** @internal */
  @Event() ldchoosefiles: EventEmitter<FileList>

  /** @internal */
  @Event() lduploadclick: EventEmitter

  private handleUploadClick = () => {
    this.lduploadclick.emit()
  }

  private handleDragEnter = (ev: DragEvent) => {
    ev.preventDefault()
    ev.stopPropagation()
    const fileList = ev.dataTransfer.files
    if (
      (this.selectMultiple || (!this.selectMultiple && fileList.length <= 1)) &&
      (this.startUpload || (!this.startUpload && !this.continueClicked))
    ) {
      this.highlighted = true
    }
  }

  private handleDragOver = (ev: DragEvent) => {
    ev.preventDefault()
    ev.stopPropagation()
    const fileList = ev.dataTransfer.files
    if (
      (this.selectMultiple || (!this.selectMultiple && fileList.length <= 1)) &&
      (this.startUpload || (!this.startUpload && !this.continueClicked))
    ) {
      this.highlighted = true
    }
  }

  private handleDragLeave = (ev: DragEvent) => {
    ev.preventDefault()
    ev.stopPropagation()
    this.highlighted = false
  }

  private handleDrop = (ev: DragEvent) => {
    ev.preventDefault()
    ev.stopPropagation()
    console.log(ev)
    this.highlighted = false

    const fileList = ev.dataTransfer.files
    console.info('fileList', fileList)

    if (
      (this.selectMultiple || (!this.selectMultiple && fileList.length <= 1)) &&
      (this.startUpload || (!this.startUpload && !this.continueClicked))
    ) {
      this.ldchoosefiles.emit(fileList)
    }
  }

  render() {
    const cl = getClassNames([
      'ld-choose-file',
      this.highlighted && 'ld-choose-file--highlighted',
    ])

    const calculateProgress = () => {
      const activeUploads = this.uploadItems.filter(
        (item) =>
          item.state == 'pending' ||
          item.state == 'paused' ||
          item.state == 'uploading' ||
          item.state == 'uploaded'
      )
      const totalSizeSum = activeUploads.reduce(
        (partialSum, file) => partialSum + file.fileSize,
        0
      )
      const uploadedSizeSum = activeUploads.reduce(
        (partialSum, file) => partialSum + file.fileSize * file.progress,
        0
      )
      const totalProgress = uploadedSizeSum / totalSizeSum / 100 || 0
      return totalProgress
    }

    return (
      <Host
        class={cl}
        onDragEnter={this.handleDragEnter}
        onDragOver={this.handleDragOver}
        onDragLeave={this.handleDragLeave}
        onDrop={this.handleDrop}
      >
        <div
          class={getClassNames([
            'ld-choose-file__content',
            this.size && `ld-choose-file--${this.size}`,
          ])}
        >
          <img
            class="ld-choose-file__image"
            src={getAssetPath('./assets/file-upload.svg')}
            alt=""
            width="142"
            height="122"
          />

          <div class="ld-choose-file__text">
            {this.startUpload ||
            (!this.startUpload && !this.continueClicked) ? (
              <Fragment>
                <ld-typo variant="h5">
                  {this.selectMultiple
                    ? 'Drag your file(s) here or browse'
                    : 'Drag your file here or browse'}
                </ld-typo>
                <ld-typo>max. 1.5 mb file size</ld-typo>
                <slot></slot>
                <ld-button onClick={this.handleUploadClick}>
                  {this.startUpload
                    ? this.selectMultiple
                      ? 'Upload file(s)'
                      : 'Upload a file'
                    : this.selectMultiple
                    ? 'Select file(s)'
                    : 'Select a file'}
                </ld-button>
              </Fragment>
            ) : (
              <Fragment>
                <ld-typo variant="h5">Upload state:</ld-typo>
                <ld-typo>
                  {
                    this.uploadItems.filter((item) => item.state == 'uploaded')
                      .length
                  }{' '}
                  of {this.uploadItems.length} file(s) uploaded.
                </ld-typo>
                <ld-typo>{calculateProgress()} % uploaded.</ld-typo>
              </Fragment>
            )}
          </div>
        </div>
      </Host>
    )
  }
}
