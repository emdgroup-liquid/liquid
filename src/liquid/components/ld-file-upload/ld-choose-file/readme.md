---
eleventyNavigation:
  key: Choose File
  parent: File Upload
layout: layout.njk
title: Choose File
permalink: components/ld-file-upload/ld-choose-file/
---

# ld-choose-file

The `ld-choose-file` component is used internally for the `ld-file-upload`. It is the visual part of the drag and drop area.

## Examples

### Default

{% example '{ "opened": true }' %}
<ld-choose-file></ld-choose-file>

<!-- React component -->

<!-- CSS component -->

{% endexample %}

<!-- Auto Generated Below -->


## Overview

TODO: emit files chosen event with file list, that's it.

## Properties

| Property      | Attribute  | Description                                              | Type                                                                                                                                              | Default     |
| ------------- | ---------- | -------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------- | ----------- |
| `key`         | `key`      | for tracking the node's identity when working with lists | `string \| number`                                                                                                                                | `undefined` |
| `maxSize`     | `max-size` | Max. file size in bytes                                  | `number`                                                                                                                                          | `1572864`   |
| `ref`         | `ref`      | reference to component                                   | `any`                                                                                                                                             | `undefined` |
| `uploadFiles` | --         | Chosen Files                                             | `{ state: "pending" \| "uploading" \| "uploaded" \| "upload failed"; fileName: string; fileSize: number; fileType: string; progress: number; }[]` | `[]`        |


## Events

| Event           | Description | Type                    |
| --------------- | ----------- | ----------------------- |
| `ldchoosefiles` |             | `CustomEvent<FileList>` |


## Shadow Parts

| Part     | Description                            |
| -------- | -------------------------------------- |
| `"list"` | `ul` element wrapping the default slot |


## Dependencies

### Used by

 - [ld-file-upload](..)

### Depends on

- [ld-typo](../../ld-typo)
- [ld-button](../../ld-button)

### Graph
```mermaid
graph TD;
  ld-choose-file --> ld-typo
  ld-choose-file --> ld-button
  ld-file-upload --> ld-choose-file
  style ld-choose-file fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
