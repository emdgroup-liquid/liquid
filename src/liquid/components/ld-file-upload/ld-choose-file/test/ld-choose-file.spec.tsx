import { newSpecPage } from '@stencil/core/testing'
import { LdChooseFile } from '../ld-choose-file'

describe('ld-choose-file', () => {
  it('renders default', async () => {
    const page = await newSpecPage({
      components: [LdChooseFile],
      html: `<ld-choose-file></ld-choose-file>`,
    })
    expect(page.root).toMatchSnapshot()
  })
})
