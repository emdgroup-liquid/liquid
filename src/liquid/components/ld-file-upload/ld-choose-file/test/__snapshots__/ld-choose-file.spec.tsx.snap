// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-choose-file renders default 1`] = `
<ld-choose-file class="ld-choose-file">
  <mock:shadow-root>
    <div class="ld-choose-file--bg ld-choose-file__content">
      <img alt="" class="ld-choose-file__image" height="122" src="/dist/liquid/assets/file-upload.svg" width="142">
      <div class="ld-choose-file__text">
        <ld-typo variant="h5">
          Drag your file(s) here or browse
        </ld-typo>
        <ld-typo>
          max. 1.5 mb file size
        </ld-typo>
        <slot></slot>
        <ld-button>
          Upload a file
        </ld-button>
      </div>
    </div>
  </mock:shadow-root>
</ld-choose-file>
`;
